src/Common/src/CoreWCF/Runtime/TaskHelpers.cs

```diff
@@ -29,120 +30,124 @@ public static async Task AsyncWait<TException>(this Task task)
         // In the EndMethod, you would use ToApmEnd<TResult> to ensure the correct exception handling
         // This will handle throwing exceptions in the correct place and ensure the IAsyncResult contains the provided
         // state object
-        public static Task<TResult> ToApm<TResult>(this Task<TResult> task, AsyncCallback callback, object state)
+        public static IAsyncResult ToApm<T>(this Task<T> task, AsyncCallback callback, object state)
+            => ToApm<T>(new ValueTask<T>(task), callback, state);
+
+        /// <summary>
+        /// Helper method to convert from Task async method to "APM" (IAsyncResult with Begin/End calls)
+        /// </summary>
+        public static IAsyncResult ToApm<T>(this ValueTask<T> task, AsyncCallback callback, object state)
         {
-            // When using APM, the returned IAsyncResult must have the passed in state object stored in AsyncState. This
-            // is so the callback can regain state. If the incoming task already holds the state object, there's no need
-            // to create a TaskCompletionSource to ensure the returned (IAsyncResult)Task has the right state object.
-            // This is a performance optimization for this special case.
-            if (task.AsyncState == state)
+            var result = new AsyncResult<T>(task, callback, state);
+            if (result.CompletedSynchronously)
```

Comment:
This will call the callback twice in the scenario where the APM method is expected to have synchronous completion optimized. In other words, I should be able to write the following:
```c#
var iar = BeginFoo(param1, callback, state);
if (iar.CompletedSynchronously)
{
    var result = EndFoo(iar);
    callback(state);
}
```
This wasn't an issue before as the old implementation also returned false for CompletedSynchronously.

Reply:
Based on my understanding i think code like that is "flawed" i.e it should not need the line "`callback(state);`" unless the user intend the callback to be executed twice.

I do belive the following is the correct usage.

``` csharp
var iar = BeginFoo(param1, callback, state);
if (iar.CompletedSynchronously)
{
    result = EndFoo(iar);
}
```
With having a corresponding check in the callback which just does nothing if CompletedSynchronously is true, just as the `Callback` method in the same file does.


This is based on [TaskToApm wrapper from the .Net Framework](https://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/TaskToApm.cs,45)  (which I found today) 
as well as the respose to [this stackoverflow question](https://stackoverflow.com/questions/1372053/asynccallback-completedsynchronously) which reference this old [indigo (wcf) blog post](https://blogs.msdn.microsoft.com/mjm/2005/05/04/asynchronous-programming-in-indigo/) 

It seems like the .Net framework version is possibly even more generic and may handle many different usages better (apart from no ValueTask support) so you could consider using that instead.

Reply:
You're right. There are a couple of different variations of the APM pattern used in .NET Framework and I was getting my wires crossed.